cms/djangoapps/modulestore_migrator/tasks.py

```diff
@@ -0,0 +1,923 @@
+from enum import Enum
+from uuid import uuid4
+
+from celery import shared_task
+from celery.utils.log import get_task_logger
+from django.db import IntegrityError
+from django.core.exceptions import ObjectDoesNotExist
+from django.utils.text import slugify
+from edx_django_utils.monitoring import set_code_owner_attribute_from_module
+from lxml import etree
+from lxml.etree import _ElementTree as XmlTree
+from opaque_keys.edx.keys import CourseKey, UsageKey
+from opaque_keys.edx.locator import (
+    CourseLocator, LibraryLocator,
+    LibraryLocatorV2, LibraryUsageLocatorV2, LibraryContainerLocator
+)
+from openedx_learning.api import authoring as authoring_api
+from openedx_learning.api.authoring_models import (
+    Collection,
+    Component,
+    ComponentVersionContent,
+    LearningPackage,
+    PublishableEntity,
+    PublishableEntityVersion,
+)
+from user_tasks.tasks import UserTask, UserTaskStatus
+
+from openedx.core.djangoapps.content_libraries.api import ContainerType, blocks
+from openedx.core.djangoapps.content_libraries import api as libraries_api
+from openedx.core.djangoapps.content_libraries.models import ContentLibrary
+from openedx.core.djangoapps.content_staging import api as staging_api
+from xmodule.modulestore import exceptions as modulestore_exceptions
+from xmodule.modulestore.django import modulestore
+
+from .constants import CONTENT_STAGING_PURPOSE_TEMPLATE
+from .data import CompositionLevel
+from .models import ModulestoreSource, ModulestoreMigration, ModulestoreBlockSource, ModulestoreBlockMigration
+
+
+log = get_task_logger(__name__)
+
+
+class MigrationStep(Enum):
+    """
+    Strings representation the state of an in-progress modulestore-to-learning-core import.
+
+    We use these values to set UserTaskStatus.state.
+    The other possible UserTaskStatus.state values are the built-in ones:
+    UserTaskStatus.{PENDING,FAILED,CANCELED,SUCCEEDED}.
+    """
+    VALIDATING_INPUT = 'Validating migration parameters'
+    CANCELLING_OLD = 'Cancelling any redundant migration tasks'
+    LOADING = 'Loading legacy content from ModulesStore'
+    STAGING = 'Staging legacy content for import'
+    PARSING = 'Parsing staged OLX'
+    IMPORTING_ASSETS = 'Importing staged files and resources'
+    IMPORTING_STRUCTURE = 'Importing staged content structure'
+    UNSTAGING = 'Cleaning staged content'
+    MAPPING_OLD_TO_NEW = 'Saving map of legacy content to migrated content'
+    FORWARDING = 'Forwarding legacy content to migrated content'
+    POPULATING_COLLECTION = 'Assigning imported items to the specified collection'
+
+
+class _MigrationTask(UserTask):
+    """
+    Base class for migrate_to_modulestore
+    """
+
+    @staticmethod
+    def calculate_total_steps(arguments_dict):
+        """
+        Get number of in-progress steps in importing process, as shown in the UI.
+        """
+        return len(list(MigrationStep))
+
+
+@dataclass(frozen=True)
+class MigrationContext:
```

Comment:
```suggestion
class _MigrationContext:
```
Let's mark this as private, as it's not meant to be used outside tasks.py.

Reply:
Agreed. I actually hesitated to mark it private and also tend to use private variables for this module.